### ISSUE TITLE
unarchive module not correctly unzipping symlinks

### ISSUE THREAD

**Issue by lukehoersten** [ansible/ansible-modules-core#1823]
##### ISSUE TYPE

Bug Report
##### COMPONENT NAME

unarchive module
##### ANSIBLE VERSION

N/A
##### SUMMARY

I have a task downloading GHC-7.10.1 from https://ghcformacosx.github.io and unarchiving it on Mac OS X 10.11 into the `/Applications/` directory. When using the `unarchive` module, symlinks are not correctly unzipped. Only a text file with the target exists in it's place. When running `unzip` on the command line directly, things work fine.

I'm running Ansible HEAD.

**Ansible Task**

``` yaml
ghc_url: https://github.com/ghcformacosx/ghc-dot-app/releases/download/v7.10.1-r0/ghc-7.10.1-r0.zip

- name: download and install ghc
  unarchive: src={{ghc_url}} dest=/Applications/ copy=no
```

**Task Result**

```
$ cd /Applications/ghc-7.10.1.app/Contents/bin
$ ls -l
total 37064
-rwxr-xr-x  1 user  staff  18874912 Mar 29 00:52 cabal
-rwxr-xr-x  1 user  staff        10 Mar 29 00:51 ghc
-rwxr-xr-x  1 user  staff       680 Mar 29 00:52 ghc-7.10.1
-rwxr-xr-x  1 user  staff        14 Mar 29 00:51 ghc-pkg
-rwxr-xr-x  1 user  staff       712 Mar 29 00:52 ghc-pkg-7.10.1
-rwxr-xr-x  1 user  staff        11 Mar 29 00:51 ghci
-rwxr-xr-x  1 user  staff       490 Mar 29 00:52 ghci-7.10.1
-rwxr-xr-x  1 user  staff        18 Mar 29 00:51 haddock
-rwxr-xr-x  1 user  staff       671 Mar 29 00:52 haddock-ghc-7.10.1
-rwxr-xr-x  1 user  staff     43776 Mar 29 00:52 hp2ps
-rwxr-xr-x  1 user  staff       642 Mar 29 00:52 hpc
-rwxr-xr-x  1 user  staff      1446 Mar 29 00:52 hsc2hs
-rwxr-xr-x  1 user  staff        13 Mar 29 00:51 runghc
-rwxr-xr-x  1 user  staff       663 Mar 29 00:52 runghc-7.10.1
-rwxr-xr-x  1 user  staff         6 Mar 29 00:51 runhaskell
```

**CLI Unzip Result**

```
$ unzip -o "/tmp/ghc-7.10.1-r0.zip" -d "/Applications/"
[...]
  inflating: /Applications/ghc-7.10.1.app/Contents/share/man/man1/ghc.1  
finishing deferred symbolic links:
  /Applications/ghc-7.10.1.app/Contents/bin/ghc -> ghc-7.10.1
  /Applications/ghc-7.10.1.app/Contents/bin/ghc-pkg -> ghc-pkg-7.10.1
  /Applications/ghc-7.10.1.app/Contents/bin/ghci -> ghci-7.10.1
  /Applications/ghc-7.10.1.app/Contents/bin/haddock -> haddock-ghc-7.10.1
  /Applications/ghc-7.10.1.app/Contents/bin/runghc -> runghc-7.10.1
/Applications/ghc-7.10.1.app/Contents/bin/runhaskell -> runghc

$ ls -l
total 37064
-rwxr-xr-x  1 user  staff  18874912 Mar 29 02:52 cabal
lrwxr-xr-x  1 user  staff        10 Jul 28 09:50 ghc -> ghc-7.10.1
-rwxr-xr-x  1 user  staff       680 Mar 29 02:52 ghc-7.10.1
lrwxr-xr-x  1 user  staff        14 Jul 28 09:50 ghc-pkg -> ghc-pkg-7.10.1
-rwxr-xr-x  1 user  staff       712 Mar 29 02:52 ghc-pkg-7.10.1
lrwxr-xr-x  1 user  staff        11 Jul 28 09:50 ghci -> ghci-7.10.1
-rwxr-xr-x  1 user  staff       490 Mar 29 02:52 ghci-7.10.1
lrwxr-xr-x  1 user  staff        18 Jul 28 09:50 haddock -> haddock-ghc-7.10.1
-rwxr-xr-x  1 user  staff       671 Mar 29 02:52 haddock-ghc-7.10.1
-rwxr-xr-x  1 user  staff     43776 Mar 29 02:52 hp2ps
-rwxr-xr-x  1 user  staff       642 Mar 29 02:52 hpc
-rwxr-xr-x  1 user  staff      1446 Mar 29 02:52 hsc2hs
lrwxr-xr-x  1 user  staff        13 Jul 28 09:50 runghc -> runghc-7.10.1
-rwxr-xr-x  1 user  staff       663 Mar 29 02:52 runghc-7.10.1
lrwxr-xr-x  1 user  staff         6 Jul 28 09:50 runhaskell -> runghc
```

**Comment by Jmainguy** [comment]
@LukeHoersten Is this still an issue with latest devel code? Digging into the code, it looks to be doing exactly what you mentioned worked on the cli.

```
cmd = '%s -o "%s" -d "%s"' % (self.cmd_path, self.src, self.dest)
```

where self.cmd is unzip, self.src is the zip, and self.dest is where its unzipping to.

**Comment by lukehoersten** [comment]
I don't use this feature anymore (since it wasn't working for me). But the code always looked right. As I mentioned, copying, pasting, and manually expanding the format string arguments and running it on the CLI worked but the code itself did not. I don't have a good setup to test at the moment unfortunately.

**Comment by ansibot** [comment]
@dagwieers, @pileofrogs, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by dagwieers** [comment]
@LukeHoersten The zip-support part has been rewritten, if you could reproduce your problem we would be grateful. If not, please close this ticket.

needs_info

**Comment by ansibot** [comment]
@LukeHoersten we are unable to validate that "" is an existing ansible module.

We are expecting a component name that matches something that can be found in:
* http://docs.ansible.com/ansible/list_of_all_modules.html

Please revise the component name in the description.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by dagwieers** [comment]
Alright, I cannot reproduce this. Works fine for me. notabug

PS What's more, this is not a very efficient way to do it because replaying the playbook causes a 177MB file download every time just to evaluate whether it is needed or not. Even in check-mode :-/
